Feat: Make it smoother to switch between nearby and explore maps

andy-ife opened the pull request:
This pr makes it easier for users to switch between the nearby and explore screens for quick comparisons e.g checking if a recommended location in nearby already has photos in explore. When switching from nearby to explore, nearby map camera data (zoom and center) is bundled and passed as fragment arguments, which are loaded in the explore map. And vice-versa.

Fixes #6100

**Changes made:**
- res/menu: extended nearby_fragment_menu with a 'Show in Explore' option, and added a new menu explore_fragment_menu with a 'Show in Nearby' option.
- main activity: extended the loadFragment() method to support passing bundles between fragments. This is how the map data is passed between nearby and explore.
- fixed memory leaks caused by main activity holding references to old fragments. 

**Tests performed**
- Tested functionality while logged in and out.
- Tested under unfavourable conditions - no internet, tapping quickly

Tested prodDebug, betaDebug on Tecno Spark 10c with API level 31.


https://github.com/user-attachments/assets/c86bf0af-f72b-4528-aa27-7000abb0d0de

nicolas-raoul commented:
The screencast looks great!

I took the liberty to add a commit because of a conflict, feel free to modify it if you think a better resolution is possible.

I am currently testing.
When switching and panning several times in areas with a lot of items (you can use the "FakeGPS" app to simulate Vienna or Tokyo for instance), I get this crash:
```
java.lang.RuntimeException: android.os.TransactionTooLargeException: data parcel size 778460 bytes
Bundle stats:
  android:viewHierarchyState [size=2708]
    android:views [size=2604]
  androidx.lifecycle.BundlableSavedStateRegistry.key [size=774884]
    android:support:activity-result [size=52704]
      KEY_COMPONENT_ACTIVITY_REGISTERED_KEYS [size=49828]
      KEY_COMPONENT_ACTIVITY_REGISTERED_RCS [size=2516]
    android:support:fragments [size=721868]
      android:support:fragments [size=721796]
PersistableBundle stats:
  [null]
at android.app.servertransaction.PendingTransactionActions$StopInfo.run(PendingTransactionActions.java:146)
at android.os.Handler.handleCallback(Handler.java:991)
at android.os.Handler.dispatchMessage(Handler.java:102)
at android.os.Looper.loopOnce(Looper.java:232)
at android.os.Looper.loop(Looper.java:317)
at android.app.ActivityThread.main(ActivityThread.java:8934)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:911)
Caused by: android.os.TransactionTooLargeException: data parcel size 778460 bytes
at android.os.BinderProxy.transactNative(Native Method)
at android.os.BinderProxy.transact(BinderProxy.java:592)
at android.app.IActivityClientController$Stub$Proxy.activityStopped(IActivityClientController.java:1504)
at android.app.ActivityClient.activityStopped(ActivityClient.java:100)
at android.app.servertransaction.PendingTransactionActions$StopInfo.run(PendingTransactionActions.java:135)
... 8 more
```
Not sure whether it is exists in the main branch or whether it is due to this branch. Any insight on this?

andy-ife commented:
Hmm... maybe a problem passing extras between fragments? I'll look into it.

nicolas-raoul commented:
Another crash, this time when coming back to the app after leaving it for a few hours. Unfortunately I am not sure how to reproduce it:
```
com.bumptech.glide.load.engine.CallbackException: Unexpected exception thrown by non-Glide code
at com.bumptech.glide.load.engine.EngineJob.callCallbackOnResourceReady(EngineJob.java:161)
at com.bumptech.glide.load.engine.EngineJob$CallResourceReady.run(EngineJob.java:428)
at android.os.Handler.handleCallback(Handler.java:991)
at android.os.Handler.dispatchMessage(Handler.java:102)
at android.os.Looper.loopOnce(Looper.java:232)
at android.os.Looper.loop(Looper.java:317)
at android.app.ActivityThread.main(ActivityThread.java:8934)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:911)
Caused by: java.lang.IllegalStateException: Fragment ExploreMapFragment{7f60aa2} (417a48e8-1927-4d74-abc2-d1a8115998c1) not attached to a context.
at androidx.fragment.app.Fragment.requireContext(Fragment.java:900)
at androidx.fragment.app.Fragment.getResources(Fragment.java:964)
at fr.free.nrw.commons.explore.map.ExploreMapFragment.addMarkerToMap(ExploreMapFragment.java:706)
at fr.free.nrw.commons.explore.map.ExploreMapFragment.addMarkersToMap(ExploreMapFragment.java:693)
at fr.free.nrw.commons.explore.map.ExploreMapPresenter.onNearbyBaseMarkerThumbsReady(ExploreMapPresenter.java:182)
at fr.free.nrw.commons.explore.map.ExploreMapController$1.onResourceReady(ExploreMapController.java:180)
at fr.free.nrw.commons.explore.map.ExploreMapController$1.onResourceReady(ExploreMapController.java:170)
at com.bumptech.glide.request.SingleRequest.onResourceReady(SingleRequest.java:639)
at com.bumptech.glide.request.SingleRequest.onResourceReady(SingleRequest.java:578)
at com.bumptech.glide.load.engine.EngineJob.callCallbackOnResourceReady(EngineJob.java:159)
```

nicolas-raoul commented:
Sorry please ignore my previous comment, it might have been a version issue on my side.

andy-ife commented:
> Another crash, this time when coming back to the app after leaving it for a few hours. Unfortunately I am not sure how to reproduce it:
> 
> ```
> com.bumptech.glide.load.engine.CallbackException: Unexpected exception thrown by non-Glide code
> at com.bumptech.glide.load.engine.EngineJob.callCallbackOnResourceReady(EngineJob.java:161)
> at com.bumptech.glide.load.engine.EngineJob$CallResourceReady.run(EngineJob.java:428)
> at android.os.Handler.handleCallback(Handler.java:991)
> at android.os.Handler.dispatchMessage(Handler.java:102)
> at android.os.Looper.loopOnce(Looper.java:232)
> at android.os.Looper.loop(Looper.java:317)
> at android.app.ActivityThread.main(ActivityThread.java:8934)
> at java.lang.reflect.Method.invoke(Native Method)
> at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
> at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:911)
> Caused by: java.lang.IllegalStateException: Fragment ExploreMapFragment{7f60aa2} (417a48e8-1927-4d74-abc2-d1a8115998c1) not attached to a context.
> at androidx.fragment.app.Fragment.requireContext(Fragment.java:900)
> at androidx.fragment.app.Fragment.getResources(Fragment.java:964)
> at fr.free.nrw.commons.explore.map.ExploreMapFragment.addMarkerToMap(ExploreMapFragment.java:706)
> at fr.free.nrw.commons.explore.map.ExploreMapFragment.addMarkersToMap(ExploreMapFragment.java:693)
> at fr.free.nrw.commons.explore.map.ExploreMapPresenter.onNearbyBaseMarkerThumbsReady(ExploreMapPresenter.java:182)
> at fr.free.nrw.commons.explore.map.ExploreMapController$1.onResourceReady(ExploreMapController.java:180)
> at fr.free.nrw.commons.explore.map.ExploreMapController$1.onResourceReady(ExploreMapController.java:170)
> at com.bumptech.glide.request.SingleRequest.onResourceReady(SingleRequest.java:639)
> at com.bumptech.glide.request.SingleRequest.onResourceReady(SingleRequest.java:578)
> at com.bumptech.glide.load.engine.EngineJob.callCallbackOnResourceReady(EngineJob.java:159)
> ```

No, you were right. All works well for the first few switches, but eventually, the app crashes with this error. I've mostly solved it though. Just testing for now. Concerning the first crash logs you sent (the TransactionTooLargeException), I haven't been able to recreate that specific error, but will look into it with further testing.

andy-ife commented:
So these are the changes I've made:
- Fixed the crash caused by the second error you sent. I've still been unable to recreate the first though.
- Refactored MainActivity to pass unit tests
- Refactored one unit test in explore: testOnCreateOptionsMenu, because now there are two possible option menus that can show in the app bar

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/explore/map/ExploreMapFragment.java:
```diff
@@ -328,25 +342,66 @@ private void performMapReadyActions() {
-     *      if bottom sheet details are expanded or collapsed hide the bottom sheet details.
+     * a) Creates bottom sheet behaviours from bottom sheet, sets initial states and visibility b)
+     * Gets the touch event on the map to perform following actions: if bottom sheet details are
+     * expanded or collapsed hide the bottom sheet details.
```
nit: would you mind reverting this reformat, as it makes the list less visible?

nicolas-raoul reviewed (changes requested):
I am not seeing changes anymore, thanks!
Just a tiny ask added.

andy-ife commented:
Changes made :+1:

nicolas-raoul reviewed (approved):
Thanks for this great contribution!
It will make many users' lives easier.
